packages/rest/src/lib/RequestManager.ts

```diff
@@ -0,0 +1,281 @@
+}
+
+/**
+ * Possible API methods to be used when doing requests
+ */
+export const enum RequestMethod {
+	Delete = 'delete',
+	Get = 'get',
+	Patch = 'patch',
+	Post = 'post',
+	Put = 'put',
+}
+
+/**
+ * Internal request options
+ *
+ * @internal
+ */
+export interface InternalRequest extends RequestData {
+	method: RequestMethod;
+	fullRoute: string;
+}
+
+/**
+ * Parsed route data for an endpoint
+ *
+ * @internal
+ */
+export interface RouteData {
+	majorParameter: string;
+	bucketRoute: string;
+	original: string;
+}
+
+/**
+ * Represents the class that manages handlers for endpoints
+ */
+export class RequestManager extends EventEmitter {
+	/**
+	 * A timeout promise that is set when we hit the global rate limit
+	 *
+	 * @default null
+	 */
+	public globalTimeout: Promise<void> | null = null;
+
+	/**
+	 * API bucket hashes that are cached from provided routes
+	 */
+	public readonly hashes = new Collection<string, string>();
+
+	/**
+	 * Request handlers created from the bucket hash and the major parameters
+	 */
+	public readonly handlers = new Collection<string, IHandler>();
+
+	// eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+	#token: string | null = null;
+
+	public constructor(public readonly options: RESTOptions) {
+		super();
+	}
+
+	public setToken(token: string) {
+		this.#token = token;
+		return this;
+	}
+
+	public async queueRequest(request: InternalRequest): Promise<unknown> {
+		// Generalize the endpoint to its route data
+		const routeID = RequestManager.generateRouteData(request.fullRoute, request.method);
+		// Get the bucket hash for the generic route, or point to a global route otherwise
+		const hash =
+			this.hashes.get(`${request.method}:${routeID.bucketRoute}`) ?? `Global(${request.method}:${routeID.bucketRoute})`;
+
+		// Get the request handler for the obtained hash, with its major parameter
+		const handler =
+			this.handlers.get(`${hash}:${routeID.majorParameter}`) ?? this.createHandler(hash, routeID.majorParameter);
+
+		// Resolve the request into usable fetch/node-fetch options
+		const { url, fetchOptions } = await this.resolveRequest(request);
+
+		// Queue the request
+		return handler.queueRequest(routeID, url, fetchOptions);
+	}
+
+	/**
+	 * Creates a new rate limit handler from a hash, based on the hash and the major parameter
+	 * @param hash The hash for the route
+	 * @param majorParameter The major parameter for this handler
+	 * @private
+	 */
+	private createHandler(hash: string, majorParameter: string) {
+		// Create the async request queue to handle requests
+		const queue = new SequentialHandler(this, hash, majorParameter);
+		// Save the queue based on its ID
+		this.handlers.set(queue.id, queue);
+
+		return queue;
+	}
+
+	/**
+	 * Formats the request data to a usable format for fetch
+	 * @param request The request data
+	 */
+	private resolveRequest(request: InternalRequest): Promise<{ url: string; fetchOptions: RequestInit }> {
```

Comment:
Why does this method return a Promise, when it's fully synchronous?

Reply:
Beecauseeee... if we throw an error, it'd be a sync error, till we get to queueRequest on the handler, where it would be an async error... and otherwise we have to disable require-await in the original function